File: sdks/go/a2ui/types_test.go
package a2ui_test

import (
	"encoding/json"
	"testing"

	"github.com/ainative/a2ui-go/a2ui"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestComponentMarshaling(t *testing.T) {
	t.Run("marshal component to JSON", func(t *testing.T) {
		component := a2ui.Component{
			Type: "button",
			ID:   "btn-1",
			Properties: map[string]interface{}{
				"label":   "Click me",
				"variant": "primary",
			},
			Children: []string{"child-1", "child-2"},
		}

		data, err := json.Marshal(component)
		require.NoError(t, err)

		var unmarshaled map[string]interface{}
		err = json.Unmarshal(data, &unmarshaled)
		require.NoError(t, err)

		assert.Equal(t, "button", unmarshaled["type"])
		assert.Equal(t, "btn-1", unmarshaled["id"])
	})

	t.Run("unmarshal component from JSON", func(t *testing.T) {
		jsonData := `{
			"type": "textField",
			"id": "field-1",
			"properties": {
				"label": "Name",
				"placeholder": "Enter your name"
			}
		}`

		var component a2ui.Component
		err := json.Unmarshal([]byte(jsonData), &component)
		require.NoError(t, err)

		assert.Equal(t, "textField", component.Type)
		assert.Equal(t, "field-1", component.ID)
		assert.Equal(t, "Name", component.Properties["label"])
	})
}

func TestCreateSurfaceMessage(t *testing.T) {
	t.Run("create valid createSurface message", func(t *testing.T) {
		msg := &a2ui.CreateSurfaceMessage{
			Type:      a2ui.MessageTypeCreateSurface,
			SurfaceID: "surface-1",
			Components: []a2ui.Component{
				{
					Type: "card",
					ID:   "card-1",
				},
			},
			DataModel: map[string]interface{}{
				"user": map[string]interface{}{
					"name": "Alice",
				},
			},
		}

		assert.Equal(t, a2ui.MessageTypeCreateSurface, msg.MessageType())
		assert.Equal(t, "surface-1", msg.SurfaceID)
		assert.Equal(t, 1, len(msg.Components))
	})

	t.Run("marshal createSurface message to JSON", func(t *testing.T) {
		msg := &a2ui.CreateSurfaceMessage{
			Type:      a2ui.MessageTypeCreateSurface,
			SurfaceID: "surface-1",
			Components: []a2ui.Component{
				{
					Type: "button",
					ID:   "btn-1",
					Properties: map[string]interface{}{
						"label": "Submit",
					},
				},
			},
			DataModel: map[string]interface{}{
				"formData": map[string]interface{}{},
			},
		}

		data, err := json.Marshal(msg)
		require.NoError(t, err)

		var unmarshaled map[string]interface{}
		err = json.Unmarshal(data, &unmarshaled)
		require.NoError(t, err)

		assert.Equal(t, "createSurface", unmarshaled["type"])
		assert.Equal(t, "surface-1", unmarshaled["surfaceId"])
		assert.NotNil(t, unmarshaled["components"])
		assert.NotNil(t, unmarshaled["dataModel"])
	})

	t.Run("unmarshal createSurface message from JSON", func(t *testing.T) {
		jsonData := `{
			"type": "createSurface",
			"surfaceId": "surface-2",
			"components": [
				{
					"type": "text",
					"id": "text-1",
					"properties": {
						"content": "Hello World"
					}
				}
			],
			"dataModel": {
				"greeting": "Hello"
			}
		}`

		var msg a2ui.CreateSurfaceMessage
		err := json.Unmarshal([]byte(jsonData), &msg)
		require.NoError(t, err)

		assert.Equal(t, a2ui.MessageTypeCreateSurface, msg.Type)
		assert.Equal(t, "surface-2", msg.SurfaceID)
		assert.Equal(t, 1, len(msg.Components))
		assert.Equal(t, "text", msg.Components[0].Type)
		assert.Equal(t, "Hello", msg.DataModel["greeting"])
	})
}

func TestUpdateComponentsMessage(t *testing.T) {
	t.Run("create update message with add operation", func(t *testing.T) {
		msg := &a2ui.UpdateComponentsMessage{
			Type:      a2ui.MessageTypeUpdateComponents,
			SurfaceID: "surface-1",
			Updates: []a2ui.ComponentUpdate{
				{
					ID:        "btn-1",
					Operation: a2ui.OperationAdd,
					Component: &a2ui.Component{
						Type: "button",
						ID:   "btn-1",
					},
				},
			},
		}

		assert.Equal(t, a2ui.MessageTypeUpdateComponents, msg.MessageType())
		assert.Equal(t, 1, len(msg.Updates))
		assert.Equal(t, a2ui.OperationAdd, msg.Updates[0].Operation)
	})

	t.Run("create update message with remove operation", func(t *testing.T) {
		msg := &a2ui.UpdateComponentsMessage{
			Type:      a2ui.MessageTypeUpdateComponents,
			SurfaceID: "surface-1",
			Updates: []a2ui.ComponentUpdate{
				{
					ID:        "btn-1",
					Operation: a2ui.OperationRemove,
				},
			},
		}

		assert.Equal(t, a2ui.OperationRemove, msg.Updates[0].Operation)
		assert.Nil(t, msg.Updates[0].Component)
	})

	t.Run("marshal and unmarshal update message", func(t *testing.T) {
		msg := &a2ui.UpdateComponentsMessage{
			Type:      a2ui.MessageTypeUpdateComponents,
			SurfaceID: "surface-1",
			Updates: []a2ui.ComponentUpdate{
				{
					ID:        "btn-1",
					Operation: a2ui.OperationUpdate,
					Component: &a2ui.Component{
						Type: "button",
						ID:   "btn-1",
						Properties: map[string]interface{}{
							"disabled": true,
						},
					},
				},
			},
		}

		data, err := json.Marshal(msg)
		require.NoError(t, err)

		var unmarshaled a2ui.UpdateComponentsMessage
		err = json.Unmarshal(data, &unmarshaled)
		require.NoError(t, err)

		assert.Equal(t, a2ui.MessageTypeUpdateComponents, unmarshaled.Type)
		assert.Equal(t, "surface-1", unmarshaled.SurfaceID)
		assert.Equal(t, 1, len(unmarshaled.Updates))
		assert.Equal(t, "btn-1", unmarshaled.Updates[0].ID)
	})
}

func TestUpdateDataModelMessage(t *testing.T) {
	t.Run("create data model update with set operation", func(t *testing.T) {
		msg := &a2ui.UpdateDataModelMessage{
			Type:      a2ui.MessageTypeUpdateDataModel,
			SurfaceID: "surface-1",
			Updates: []a2ui.DataUpdate{
				{
					Path:      "/user/name",
					Operation: a2ui.DataOperationSet,
					Value:     "Alice",
				},
			},
		}

		assert.Equal(t, a2ui.MessageTypeUpdateDataModel, msg.MessageType())
		assert.Equal(t, 1, len(msg.Updates))
		assert.Equal(t, "/user/name", msg.Updates[0].Path)
		assert.Equal(t, "Alice", msg.Updates[0].Value)
	})

	t.Run("create data model update with remove operation", func(t *testing.T) {
		msg := &a2ui.UpdateDataModelMessage{
			Type:      a2ui.MessageTypeUpdateDataModel,
			SurfaceID: "surface-1",
			Updates: []a2ui.DataUpdate{
				{
					Path:      "/user/age",
					Operation: a2ui.DataOperationRemove,
				},
			},
		}

		assert.Equal(t, a2ui.DataOperationRemove, msg.Updates[0].Operation)
		assert.Nil(t, msg.Updates[0].Value)
	})

	t.Run("marshal and unmarshal data model update", func(t *testing.T) {
		msg := &a2ui.UpdateDataModelMessage{
			Type:      a2ui.MessageTypeUpdateDataModel,
			SurfaceID: "surface-1",
			Updates: []a2ui.DataUpdate{
				{
					Path:      "/items/-",
					Operation: a2ui.DataOperationSet,
					Value:     "new item",
				},
			},
		}

		data, err := json.Marshal(msg)
		require.NoError(t, err)

		var unmarshaled a2ui.UpdateDataModelMessage
		err = json.Unmarshal(data, &unmarshaled)
		require.NoError(t, err)

		assert.Equal(t, "/items/-", unmarshaled.Updates[0].Path)
		assert.Equal(t, "new item", unmarshaled.Updates[0].Value)
	})
}

func TestUserActionMessage(t *testing.T) {
	t.Run("create user action message", func(t *testing.T) {
		msg := &a2ui.UserActionMessage{
			Type:        a2ui.MessageTypeUserAction,
			SurfaceID:   "surface-1",
			Action:      "submit",
			ComponentID: "btn-1",
			Context: map[string]interface{}{
				"formData": map[string]interface{}{
					"name": "Alice",
				},
			},
		}

		assert.Equal(t, a2ui.MessageTypeUserAction, msg.MessageType())
		assert.Equal(t, "submit", msg.Action)
		assert.Equal(t, "btn-1", msg.ComponentID)
	})

	t.Run("marshal and unmarshal user action", func(t *testing.T) {
		msg := &a2ui.UserActionMessage{
			Type:        a2ui.MessageTypeUserAction,
			SurfaceID:   "surface-1",
			Action:      "click",
			ComponentID: "btn-submit",
			Context: map[string]interface{}{
				"timestamp": 1234567890,
			},
			DataModel: map[string]interface{}{
				"currentUser": "Alice",
			},
		}

		data, err := json.Marshal(msg)
		require.NoError(t, err)

		var unmarshaled a2ui.UserActionMessage
		err = json.Unmarshal(data, &unmarshaled)
		require.NoError(t, err)

		assert.Equal(t, "click", unmarshaled.Action)
		assert.Equal(t, "btn-submit", unmarshaled.ComponentID)
		assert.NotNil(t, unmarshaled.Context)
		assert.NotNil(t, unmarshaled.DataModel)
	})
}

func TestErrorMessage(t *testing.T) {
	t.Run("create error message", func(t *testing.T) {
		msg := &a2ui.ErrorMessage{
			Type:    a2ui.MessageTypeError,
			Code:    "VALIDATION_ERROR",
			Message: "Invalid input",
			Details: map[string]interface{}{
				"field":  "email",
				"reason": "invalid format",
			},
		}

		assert.Equal(t, a2ui.MessageTypeError, msg.MessageType())
		assert.Equal(t, "VALIDATION_ERROR", msg.Code)
		assert.Equal(t, "Invalid input", msg.Message)
	})

	t.Run("marshal and unmarshal error message", func(t *testing.T) {
		msg := &a2ui.ErrorMessage{
			Type:    a2ui.MessageTypeError,
			Code:    "AUTH_ERROR",
			Message: "Unauthorized",
		}

		data, err := json.Marshal(msg)
		require.NoError(t, err)

		var unmarshaled a2ui.ErrorMessage
		err = json.Unmarshal(data, &unmarshaled)
		require.NoError(t, err)

		assert.Equal(t, "AUTH_ERROR", unmarshaled.Code)
		assert.Equal(t, "Unauthorized", unmarshaled.Message)
	})
}

func TestPingPongMessages(t *testing.T) {
	t.Run("create ping message", func(t *testing.T) {
		msg := &a2ui.PingMessage{
			Type: a2ui.MessageTypePing,
		}

		assert.Equal(t, a2ui.MessageTypePing, msg.MessageType())
	})

	t.Run("create pong message", func(t *testing.T) {
		msg := &a2ui.PongMessage{
			Type: a2ui.MessageTypePong,
		}

		assert.Equal(t, a2ui.MessageTypePong, msg.MessageType())
	})

	t.Run("marshal and unmarshal ping/pong", func(t *testing.T) {
		ping := &a2ui.PingMessage{Type: a2ui.MessageTypePing}
		pong := &a2ui.PongMessage{Type: a2ui.MessageTypePong}

		pingData, err := json.Marshal(ping)
		require.NoError(t, err)

		pongData, err := json.Marshal(pong)
		require.NoError(t, err)

		var unmarshaledPing a2ui.PingMessage
		err = json.Unmarshal(pingData, &unmarshaledPing)
		require.NoError(t, err)
		assert.Equal(t, a2ui.MessageTypePing, unmarshaledPing.Type)

		var unmarshaledPong a2ui.PongMessage
		err = json.Unmarshal(pongData, &unmarshaledPong)
		require.NoError(t, err)
		assert.Equal(t, a2ui.MessageTypePong, unmarshaledPong.Type)
	})
}

func TestDeleteSurfaceMessage(t *testing.T) {
	t.Run("create delete surface message", func(t *testing.T) {
		msg := &a2ui.DeleteSurfaceMessage{
			Type:      a2ui.MessageTypeDeleteSurface,
			SurfaceID: "surface-1",
		}

		assert.Equal(t, a2ui.MessageTypeDeleteSurface, msg.MessageType())
		assert.Equal(t, "surface-1", msg.SurfaceID)
	})

	t.Run("marshal and unmarshal delete surface", func(t *testing.T) {
		msg := &a2ui.DeleteSurfaceMessage{
			Type:      a2ui.MessageTypeDeleteSurface,
			SurfaceID: "surface-2",
		}

		data, err := json.Marshal(msg)
		require.NoError(t, err)

		var unmarshaled a2ui.DeleteSurfaceMessage
		err = json.Unmarshal(data, &unmarshaled)
		require.NoError(t, err)

		assert.Equal(t, "surface-2", unmarshaled.SurfaceID)
	})
}

File: sdks/rust/a2ui-sdk/tests/types_test.rs
/// A2UI Type Definitions Tests
///
/// Comprehensive tests for all A2UI protocol message types.
use a2ui_sdk::*;
use serde_json::json;
use std::collections::HashMap;

#[test]
fn test_update_components_message() {
    let msg = UpdateComponentsMessage {
        message_type: MessageType::UpdateComponents,
        surface_id: "test-surface".to_string(),
        updates: vec![ComponentUpdate {
            id: "btn-1".to_string(),
            operation: Operation::Update,
            component: Some(Component {
                component_type: "button".to_string(),
                id: "btn-1".to_string(),
                properties: HashMap::new(),
                children: vec![],
            }),
        }],
    };

    assert_eq!(msg.message_type(), MessageType::UpdateComponents);

    let json = serde_json::to_value(&msg).unwrap();
    assert_eq!(json["type"], "updateComponents");
    assert_eq!(json["surfaceId"], "test-surface");
}

#[test]
fn test_update_data_model_message() {
    let msg = UpdateDataModelMessage {
        message_type: MessageType::UpdateDataModel,
        surface_id: "test-surface".to_string(),
        updates: vec![DataUpdate {
            path: "/user/name".to_string(),
            operation: DataOperation::Set,
            value: Some(json!("Alice")),
        }],
    };

    assert_eq!(msg.message_type(), MessageType::UpdateDataModel);

    let json = serde_json::to_value(&msg).unwrap();
    assert_eq!(json["type"], "updateDataModel");
}

#[test]
fn test_user_action_message() {
    let msg = UserActionMessage {
        message_type: MessageType::UserAction,
        surface_id: "test-surface".to_string(),
        action: "submit".to_string(),
        component_id: Some("btn-1".to_string()),
        context: HashMap::new(),
    };

    assert_eq!(msg.message_type(), MessageType::UserAction);

    let json = serde_json::to_value(&msg).unwrap();
    assert_eq!(json["type"], "userAction");
    assert_eq!(json["action"], "submit");
}

#[test]
fn test_error_message() {
    let msg = ErrorMessage {
        message_type: MessageType::Error,
        code: "VALIDATION_ERROR".to_string(),
        message: "Invalid input".to_string(),
        details: HashMap::new(),
    };

    assert_eq!(msg.message_type(), MessageType::Error);

    let json = serde_json::to_value(&msg).unwrap();
    assert_eq!(json["type"], "error");
    assert_eq!(json["code"], "VALIDATION_ERROR");
}

#[test]
fn test_delete_surface_message() {
    let msg = DeleteSurfaceMessage {
        message_type: MessageType::DeleteSurface,
        surface_id: "test-surface".to_string(),
    };

    assert_eq!(msg.message_type(), MessageType::DeleteSurface);

    let json = serde_json::to_value(&msg).unwrap();
    assert_eq!(json["type"], "deleteSurface");
}

#[test]
fn test_ping_message() {
    let msg = PingMessage {
        message_type: MessageType::Ping,
    };

    assert_eq!(msg.message_type(), MessageType::Ping);

    let json = serde_json::to_value(&msg).unwrap();
    assert_eq!(json["type"], "ping");
}

#[test]
fn test_pong_message() {
    let msg = PongMessage {
        message_type: MessageType::Pong,
    };

    assert_eq!(msg.message_type(), MessageType::Pong);

    let json = serde_json::to_value(&msg).unwrap();
    assert_eq!(json["type"], "pong");
}

#[test]
fn test_operation_serialization() {
    let op = Operation::Add;
    let json = serde_json::to_string(&op).unwrap();
    assert_eq!(json, "\"add\"");

    let op = Operation::Update;
    let json = serde_json::to_string(&op).unwrap();
    assert_eq!(json, "\"update\"");

    let op = Operation::Remove;
    let json = serde_json::to_string(&op).unwrap();
    assert_eq!(json, "\"remove\"");
}

#[test]
fn test_data_operation_serialization() {
    let op = DataOperation::Set;
    let json = serde_json::to_string(&op).unwrap();
    assert_eq!(json, "\"set\"");

    let op = DataOperation::Remove;
    let json = serde_json::to_string(&op).unwrap();
    assert_eq!(json, "\"remove\"");
}

#[test]
fn test_message_type_deserialization() {
    let json = "\"createSurface\"";
    let msg_type: MessageType = serde_json::from_str(json).unwrap();
    assert_eq!(msg_type, MessageType::CreateSurface);

    let json = "\"updateComponents\"";
    let msg_type: MessageType = serde_json::from_str(json).unwrap();
    assert_eq!(msg_type, MessageType::UpdateComponents);

    let json = "\"ping\"";
    let msg_type: MessageType = serde_json::from_str(json).unwrap();
    assert_eq!(msg_type, MessageType::Ping);
}

#[test]
fn test_all_message_types_equality() {
    assert_eq!(MessageType::CreateSurface, MessageType::CreateSurface);
    assert_eq!(MessageType::UpdateComponents, MessageType::UpdateComponents);
    assert_eq!(MessageType::UpdateDataModel, MessageType::UpdateDataModel);
    assert_eq!(MessageType::UserAction, MessageType::UserAction);
    assert_eq!(MessageType::Error, MessageType::Error);
    assert_eq!(MessageType::DeleteSurface, MessageType::DeleteSurface);
    assert_eq!(MessageType::Ping, MessageType::Ping);
    assert_eq!(MessageType::Pong, MessageType::Pong);
}

#[test]
fn test_operation_equality() {
    assert_eq!(Operation::Add, Operation::Add);
    assert_eq!(Operation::Update, Operation::Update);
    assert_eq!(Operation::Remove, Operation::Remove);
}

#[test]
fn test_data_operation_equality() {
    assert_eq!(DataOperation::Set, DataOperation::Set);
    assert_eq!(DataOperation::Remove, DataOperation::Remove);
}

#[test]
fn test_component_with_children() {
    let component = Component {
        component_type: "container".to_string(),
        id: "container-1".to_string(),
        properties: HashMap::new(),
        children: vec![
            Component {
                component_type: "button".to_string(),
                id: "btn-1".to_string(),
                properties: HashMap::new(),
                children: vec![],
            }
        ],
    };

    let json = serde_json::to_value(&component).unwrap();
    assert_eq!(json["type"], "container");
    assert!(json["children"].is_array());
    assert_eq!(json["children"][0]["type"], "button");
}

#[test]
fn test_complete_create_surface_roundtrip() {
    let original = CreateSurfaceMessage {
        message_type: MessageType::CreateSurface,
        surface_id: "test".to_string(),
        components: vec![
            Component {
                component_type: "text".to_string(),
                id: "text-1".to_string(),
                properties: {
                    let mut props = HashMap::new();
                    props.insert("value".to_string(), json!("Hello"));
                    props
                },
                children: vec![],
            }
        ],
        data_model: {
            let mut model = HashMap::new();
            model.insert("count".to_string(), json!(0));
            model
        },
    };

    let json = serde_json::to_string(&original).unwrap();
    let deserialized: CreateSurfaceMessage = serde_json::from_str(&json).unwrap();

    assert_eq!(deserialized.surface_id, "test");
    assert_eq!(deserialized.components.len(), 1);
    assert_eq!(deserialized.data_model.get("count").unwrap(), &json!(0));
}
